policy-test/src/curl.rs

```diff
@@ -0,0 +1,233 @@
+                annotations: Some(convert_args!(btreemap!(
+                    "linkerd.io/inject" => inject.to_string(),
+                    "config.linkerd.io/proxy-log-level" => "linkerd=trace,info",
+                ))),
+                ..Default::default()
+            },
+            spec: Some(k8s::PodSpec {
+                service_account: Some("curl".to_string()),
+                init_containers: Some(vec![k8s::api::core::v1::Container {
+                    name: "wait-for-nginx".to_string(),
+                    image: Some("docker.io/bitnami/kubectl:latest".to_string()),
+                    args: Some(
+                        vec![
+                            "wait",
+                            "--timeout=60s",
+                            "--for=delete",
+                            "--namespace",
+                            ns,
+                            "cm",
+                            "curl-lock",
+                        ]
+                        .into_iter()
+                        .map(Into::into)
+                        .collect(),
+                    ),
+                    ..Default::default()
+                }]),
+                containers: vec![k8s::api::core::v1::Container {
+                    name: "curl".to_string(),
+                    image: Some("docker.io/curlimages/curl:latest".to_string()),
+                    args: Some(
+                        vec!["curl", "-sSfv", target_url]
+                            .into_iter()
+                            .map(Into::into)
+                            .collect(),
+                    ),
+                    ..Default::default()
+                }],
+                restart_policy: Some("Never".to_string()),
+                ..Default::default()
+            }),
+            ..k8s::Pod::default()
+        }
+    }
+}
+
+impl Running {
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    pub async fn ip(&self) -> std::net::IpAddr {
+        super::await_pod_ip(&self.client, &self.namespace, &self.name).await
+    }
+
+    pub async fn exit_code(self) -> i32 {
+        fn get_exit_code(pod: &k8s::Pod) -> Option<i32> {
+            if let Some(status) = &pod.status {
+                if let Some(containers) = &status.container_statuses {
+                    for c in containers.iter() {
+                        if c.name == "curl" {
+                            if let Some(state) = &c.state {
+                                if let Some(terminated) = &state.terminated {
+                                    tracing::debug!(ns = %pod.namespace().unwrap(), pod = %pod.name(), "Curl exited");
+                                    return Some(terminated.exit_code);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            tracing::debug!(ns = %pod.namespace().unwrap(), pod = %pod.name(), "Curl has not exited");
+            None
```

Comment:
nit, take it or leave it: i feel like we could maybe avoid the pyramid of death here if we used `Option`'s combinators:
```suggestion
            let containers = &pod.status
                .and_then(|status| status.containers_statuses.map(|containers| containers.iter())
                .iter()
                .flatten();
            for c in containers {
                if c.name == "curl" {
                    if let Some(terminated) = &c.state.and_then(|state| state.terminated) {
                       tracing::debug!(ns = %pod.namespace().unwrap(), pod = %pod.name(), "Curl exited");
                       return Some(terminated.exit_code);
                     }
                 }
            }
            tracing::debug!(ns = %pod.namespace().unwrap(), pod = %pod.name(), "Curl has not exited");
            None
```

alternatively, we could use `?` instead of the `and_then`/`Option::iter`/`flatten` thing, but then we wouldn't emit the `tracing` event at the end of the function.